File: threading/python/03_thread_safe_singleton.py
"""
QUESTION 3: Thread-Safe Singleton Pattern (Databricks/Glean)
=============================================================

Problem Statement:
------------------
Implement a thread-safe singleton pattern that ensures only one instance is created
even when multiple threads try to create instances simultaneously.

Companies: Common at all three - used for database connections, config managers, caches

Key Concepts:
- Double-checked locking
- Memory barriers
- Class-level locking
- Python's GIL consideration
"""

import threading
import time


# ============================================================================
# SOLUTION 1: Double-Checked Locking (Classic Interview Answer)
# ============================================================================

class SingletonDoubleChecked:
    """
    Double-Checked Locking Singleton
    
    EXPLANATION:
    ------------
    1. First check: Avoid acquiring lock if instance exists (performance)
    2. Lock acquisition: Ensures only one thread creates instance
    3. Second check: Another thread might have created instance while we waited
    
    WHY DOUBLE CHECK?
    - Single check with lock: Every call acquires lock (slow)
    - Single check without lock: Race condition (two instances)
    - Double check: Fast path when instance exists, safe when creating
    
    PYTHON NOTE: Python's GIL makes this somewhat redundant, but this pattern
    is critical for Java/C++ interviews!
    """
    
    _instance = None
    _lock = threading.Lock()
    
    def __new__(cls):
        # First check (no locking)
        if cls._instance is None:
            with cls._lock:
                # Second check (with locking)
                if cls._instance is None:
                    print(f"Creating singleton instance in thread {threading.current_thread().name}")
                    cls._instance = super().__new__(cls)
                    cls._instance._initialized = False
        return cls._instance
    
    def __init__(self):
        if not self._initialized:
            self.value = 0
            self._initialized = True


# ============================================================================
# SOLUTION 2: Using Metaclass (Pythonic Way)
# ============================================================================

class SingletonMeta(type):
    """
    Thread-safe Singleton Metaclass
    
    EXPLANATION:
    ------------
    Metaclass controls class creation. __call__ is invoked when you do ClassName().
    This approach is cleaner and more Pythonic.
    """
    
    _instances = {}
    _lock = threading.Lock()
    
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            with cls._lock:
                if cls not in cls._instances:
                    instance = super().__call__(*args, **kwargs)
                    cls._instances[cls] = instance
        return cls._instances[cls]


class DatabaseConnection(metaclass=SingletonMeta):
    """Example singleton using metaclass"""
    
    def __init__(self):
        self.connection_id = id(self)
        print(f"Database connection created: {self.connection_id}")


# ============================================================================
# SOLUTION 3: Module-Level (Simplest Pythonic Way)
# ============================================================================

class _Config:
    """Private config class"""
    def __init__(self):
        self.settings = {}

# Module-level singleton - Python imports are thread-safe
config = _Config()


# ============================================================================
# SOLUTION 4: Using threading.Lock with decorator
# ============================================================================

def synchronized(lock):
    """Decorator for thread-safe methods"""
    def decorator(func):
        def wrapper(*args, **kwargs):
            with lock:
                return func(*args, **kwargs)
        return wrapper
    return decorator


class SingletonDecorator:
    _instance = None
    _lock = threading.Lock()
    
    @classmethod
    @synchronized(_lock)
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance


def test_singleton():
    """Test that singleton works correctly with multiple threads"""
    instances = []
    
    def create_instance():
        instance = SingletonDoubleChecked()
        instances.append(id(instance))
        time.sleep(0.01)
    
    threads = [threading.Thread(target=create_instance) for _ in range(10)]
    for t in threads:
        t.start()
    for t in threads:
        t.join()
    
    # All instances should have the same id
    assert len(set(instances)) == 1, "Singleton violated!"
    print(f"Success! All {len(instances)} instances are the same: {instances[0]}")


if __name__ == "__main__":
    test_singleton()

File: threading/cpp/03_thread_safe_singleton.cpp
/*
 * QUESTION 3: Thread-Safe Singleton Pattern (Databricks/Glean)
 * =============================================================
 * 
 * Problem: Ensure only one instance is created even with concurrent access.
 * 
 * Key Concepts: Double-checked locking, std::call_once, Meyer's Singleton
 * 
 * Compile: g++ -std=c++17 -pthread 03_thread_safe_singleton.cpp -o singleton
 */

#include <iostream>
#include <thread>
#include <mutex>
#include <vector>
#include <atomic>

// SOLUTION 1: Double-Checked Locking Pattern (DCLP)
class SingletonDCLP {
    /*
     * EXPLANATION:
     * 1. First check (no lock): Fast path when instance exists
     * 2. Lock acquisition: Only one thread creates
     * 3. Second check: Another thread might have created while waiting
     * 
     * CRITICAL: Use std::atomic for instance_ to prevent memory reordering!
     * Without atomic, compiler might reorder writes (allocate -> assign -> construct)
     */
private:
    static std::atomic<SingletonDCLP*> instance_;
    static std::mutex mutex_;
    int value_ = 0;

    SingletonDCLP() {
        std::cout << "SingletonDCLP created by thread " 
                  << std::this_thread::get_id() << std::endl;
    }

public:
    static SingletonDCLP* getInstance() {
        SingletonDCLP* tmp = instance_.load(std::memory_order_acquire);
        if (tmp == nullptr) {
            std::lock_guard<std::mutex> lock(mutex_);
            tmp = instance_.load(std::memory_order_relaxed);
            if (tmp == nullptr) {
                tmp = new SingletonDCLP();
                instance_.store(tmp, std::memory_order_release);
            }
        }
        return tmp;
    }

    void setValue(int v) { value_ = v; }
    int getValue() const { return value_; }
};

std::atomic<SingletonDCLP*> SingletonDCLP::instance_{nullptr};
std::mutex SingletonDCLP::mutex_;


// SOLUTION 2: std::call_once (Recommended)
class SingletonCallOnce {
    /*
     * EXPLANATION:
     * std::call_once guarantees the callable is executed exactly once.
     * Thread-safe by standard. Simple and clean!
     */
private:
    static std::unique_ptr<SingletonCallOnce> instance_;
    static std::once_flag once_flag_;

    SingletonCallOnce() {
        std::cout << "SingletonCallOnce created by thread "
                  << std::this_thread::get_id() << std::endl;
    }

public:
    static SingletonCallOnce* getInstance() {
        std::call_once(once_flag_, []() {
            instance_.reset(new SingletonCallOnce());
        });
        return instance_.get();
    }
};

std::unique_ptr<SingletonCallOnce> SingletonCallOnce::instance_;
std::once_flag SingletonCallOnce::once_flag_;


// SOLUTION 3: Meyer's Singleton (C++11 Guaranteed Thread-Safe)
class SingletonMeyer {
    /*
     * EXPLANATION:
     * C++11 guarantees static local variable initialization is thread-safe.
     * The simplest and most elegant solution!
     * 
     * "If control enters the declaration concurrently while the variable
     *  is being initialized, the concurrent execution shall wait."
     */
private:
    SingletonMeyer() {
        std::cout << "SingletonMeyer created by thread "
                  << std::this_thread::get_id() << std::endl;
    }

public:
    static SingletonMeyer& getInstance() {
        static SingletonMeyer instance;  // Thread-safe in C++11!
        return instance;
    }

    SingletonMeyer(const SingletonMeyer&) = delete;
    SingletonMeyer& operator=(const SingletonMeyer&) = delete;
};


void testSingleton() {
    // Test Meyer's Singleton
    auto* s1 = &SingletonMeyer::getInstance();
    auto* s2 = &SingletonMeyer::getInstance();
    std::cout << "Same instance: " << (s1 == s2 ? "YES" : "NO") << std::endl;
}

int main() {
    std::vector<std::thread> threads;
    
    // Multiple threads try to get singleton
    for (int i = 0; i < 10; ++i) {
        threads.emplace_back(testSingleton);
    }
    
    for (auto& t : threads) t.join();
    
    std::cout << "All tests passed!" << std::endl;
    return 0;
}

